ksql-engine/src/main/java/io/confluent/ksql/engine/TopicAccessValidator.java

```diff
@@ -0,0 +1,148 @@
+   * and persistent query statements.
+   * </p>
+   * It checks for Read access to all the source topics and Write access to the target topic.
+   * </p>
+   * Other permissions checks are already validated by the {@code Injectors} which attempts to
+   * create and/or delete topics using the same {@code ServiceContext}.
+   *
+   * @param statement The statement to verify for permissions.
+   * @throws KsqlException If a topic is not authorized for access, or the topic does not exist.
+   */
+  public void validate(final Statement statement) {
+    if (statement instanceof Query) {
+      validateQueryTopicSources((Query)statement);
+    } else if (statement instanceof InsertInto) {
+      validateInsertInto((InsertInto)statement);
+    } else if (statement instanceof CreateAsSelect) {
+      validateCreateAsSelect((CreateAsSelect)statement);
+    }
+  }
+
+  private void validateQueryTopicSources(final Query query) {
+    sourceTopicsExtractor.process(query, null);
+    for (String kafkaTopic : sourceTopicsExtractor.getSourceTopics()) {
+      checkAccess(kafkaTopic, AclOperation.READ);
+    }
+  }
+
+  private void validateCreateAsSelect(final CreateAsSelect createAsSelect) {
+    /*
+     * Check topic access for CREATE STREAM/TABLE AS SELECT statements.
+     *
+     * Validates Write on the target topic if exists, and Read on the query sources topics.
+     *
+     * The Create access is validated by the TopicCreateInjector which will attempt to create
+     * the target topic using the same ServiceContext used for validation.
+     */
+
+    validateQueryTopicSources(createAsSelect.getQuery());
+
+    // At this point, the topic should have been created by the TopicCreateInjector
+    final String kafkaTopic = getSourceTopicFromCreateAsSelect(createAsSelect);
+    checkAccess(kafkaTopic, AclOperation.WRITE);
+  }
+
+  private void validateInsertInto(final InsertInto insertInto) {
+    /*
+     * Check topic access for INSERT INTO statements.
+     *
+     * Validates Write on the target topic, and Read on the query sources topics.
+     */
+
+    validateQueryTopicSources(insertInto.getQuery());
+
+    final String kafkaTopic = getSourceTopicName(insertInto.getTarget().getSuffix());
+    checkAccess(kafkaTopic, AclOperation.WRITE);
+  }
+
+  private String getSourceTopicName(final String streamOrTable) {
+    final DataSource<?> dataSource = metaStore.getSource(streamOrTable);
+    if (dataSource == null) {
+      throw new KsqlException("Cannot validate for topic access from an unknown stream/table: "
+          + streamOrTable);
+    }
+
+    return dataSource.getKafkaTopicName();
+  }
+
+  /**
+   * Checks if the ServiceContext has access to the topic with the specified AclOperation.
+   */
+  private void checkAccess(final String topicName, final AclOperation operation) {
+    final Set<AclOperation> authorizedOperations = serviceContext.getTopicClient()
```

Comment:
What if the user's Kafka cluster doesn't use ACLs at all? Won't this check always fail?

Reply:
mmmm, that's interesting. I didn't check this until you mentioned it. When ACLs are not enabled, the operations list is empty. This causes the KSQL command to fail with no access to the topic.

I think the only way to fix it is to add a new KSQL configuration that enables ACL checks. I will add one, what do you think?